computed/index.test.ts

```diff
@@ -64,26 +70,136 @@ test('works with single store', () => {
   unbind()
 })
 
-test('prevents diamond dependency problem', () => {
+let replacer =
+  (...args: [string, string]) =>
+  (v: string) =>
+    v.replace(...args)
+
+test('prevents diamond dependency problem 1', () => {
+  let store = atom<number>(0)
+  let values: string[] = []
+
+  let a = computed(store, v => `a${v}`)
+  let b = computed(a, replacer('a', 'b'))
+  let c = computed(a, replacer('a', 'c'))
+  let d = computed(a, replacer('a', 'd'))
+
+  let combined = computed([b, c, d], ($b, $c, $d) => `${$b}${$c}${$d}`)
+
+  let unsubscribe = combined.subscribe(v => {
+    values.push(v)
+  })
+
+  equal(values, ['b0c0d0'])
+
+  store.set(1)
+  store.set(2)
+
+  equal(values, ['b0c0d0', 'b1c1d1', 'b2c2d2'])
+
+  unsubscribe()
+})
+
+test('prevents diamond dependency problem 2', () => {
+  let store = atom<number>(0)
+  let values: string[] = []
+
+  let a = computed(store, v => `a${v}`)
+  let b = computed(a, replacer('a', 'b'))
+  let c = computed(b, replacer('b', 'c'))
+  let d = computed(c, replacer('c', 'd'))
+  let e = computed(d, replacer('d', 'e'))
+
+  let combined = computed([a, e], (...args) => args.join(''))
+
+  let unsubscribe = combined.subscribe(v => {
+    values.push(v)
+  })
+
+  equal(values, ['a0e0'])
+
+  store.set(1)
+  equal(values, ['a0e0', 'a1e1'])
+
+  unsubscribe()
+})
+
+test('prevents diamond dependency problem 3', () => {
   let store = atom<number>(0)
   let values: string[] = []
 
-  let a = computed(store, count => `a${count}`)
-  let b = computed(store, count => `b${count}`)
-  let combined = computed([a, b], (first, second) => `${first}${second}`)
+  let a = computed(store, $store => `a${$store}`)
+  let b = computed(a, replacer('a', 'b'))
+  let c = computed(b, replacer('b', 'c'))
+  let d = computed(c, replacer('c', 'd'))
+
+  let combined = computed(
+    [a, b, c, d],
+    ($a, $b, $c, $d) => `${$a}${$b}${$c}${$d}`
+  )
 
   let unsubscribe = combined.subscribe(v => {
     values.push(v)
   })
 
-  equal(values, ['a0b0'])
+  equal(values, ['a0b0c0d0'])
 
   store.set(1)
-  equal(values, ['a0b0', 'a1b1'])
+  equal(values, ['a0b0c0d0', 'a1b1c1d1'])
 
   unsubscribe()
 })
 
+test('prevents diamond dependency problem 4 (complex)', () => {
```

Comment:
Should I add a scheme for the data flow is this test? I think without a scheme nobody can understand it.